src/xbot/common/command/scripted/ScriptedCommand.java

```diff
@@ -0,0 +1,164 @@
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.log4j.Logger;
+import org.apache.log4j.Priority;
+
+import edu.wpi.first.wpilibj.command.Command;
+import edu.wpi.first.wpilibj.command.Scheduler;
+
+/**
+ * A command to facilitate execution of a JavaScript script with access
+ * to robot functionality such as Command invocation.
+ *
+ */
+public class ScriptedCommand extends Command {
+
+    static Logger log = Logger.getLogger(ScriptedCommand.class);
+    
+    private ScriptedCommandThread execThread;
+
+    ScriptedCommandFactory availableCommandFactory;
+    
+    File scriptFile = null;
+    String manualScriptText = null, manualScriptName = null;
+    
+    /**
+     * Contains all the commands that this script has invoked.
+     * NOTE: Not all commands in this collection are guaranteed to be running;
+     * being in this list only signifies that a command was invoked at some
+     * point by this script.
+     */
+    private Set<Command> invokedCommands;
+    
+    private volatile Set<ScriptedCommandCompletionLock> completionLocks;
+    
+    public ScriptedCommand(File scriptFile, ScriptedCommandFactory availableCommandFactory) {
+        this.availableCommandFactory = availableCommandFactory;
+        
+        this.scriptFile = scriptFile;
+        
+        log.debug("Instantiated with file path\"" + scriptFile.getAbsolutePath() + "\"");
+    }
+    
+    public ScriptedCommand(String manualScriptText, String manualScriptName, ScriptedCommandFactory availableCommandFactory) {
+        this.availableCommandFactory = availableCommandFactory;
+        
+        this.manualScriptText = manualScriptText;
+        this.manualScriptName = manualScriptName;
+        
+        log.debug("Instantiated with script string");
+    }
+    
+    private void initializeExecThread() {
+        log.debug("Initializing exec thread");
+        
+        if(scriptFile == null) {
+            log.debug("Using manual script text for exec");
+            execThread = new ScriptedCommandThread(manualScriptText, manualScriptName, this, availableCommandFactory);
+        }
+        else {
+            log.debug("Using file path for exec");
+            execThread = new ScriptedCommandThread(scriptFile, this, availableCommandFactory);
+        }
+        
+        execThread.setUncaughtExceptionHandler(new UncaughtExceptionHandler() {
+            
+            @Override
+            public void uncaughtException(Thread t, Throwable e) {
+                Logger log = Logger.getLogger(ScriptedCommandThread.class);
+                log.error("Uncaught exception in autonomous thread! " + e.toString());
+            }
+        });
+    }
+    
+    /**
+     * For internal use only!
+     * @param command
+     */
+    public synchronized void invokeCommand(Command command) {
+        log.info("Adding command " + command + " to scheduler");
+        
+        Scheduler.getInstance().add(command);
```

Comment:
Injection in this case is more code for no extra functionality, and I was pretty lazy :wink:... I'll change this to make it uniform *JIC*.